sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala

```diff
@@ -93,6 +93,21 @@ object SubqueryExpression {
     }.isDefined
   }
 
+  /**
+   * Returns true when an expression contains a correlated IN or correlated EXISTS and
+   * false otherwise. An IN is non-correlated only if the left values are foldable
+   * and the subquery has no outer references.
+   */
+  def hasCorrelatedInOrExists(e: Expression): Boolean = {
+    e.find {
+      case InSubquery(values, ListQuery(_, children, _, _)) =>
+        values.exists(!_.foldable) || children.nonEmpty
+      case Exists(_, children, _) =>
+        children.nonEmpty
+      case _ => false
+    }.isDefined
+  }
+
```

Comment:
@francis0407 Like wenchen, i can see a source of confusion here. So if we have a query like 
select * from t1 where c1 in (1, 2), we will flag is at correlated, right ? I understand what you are doing here :-). But, future caller of this function may not understand the subtle difference (since its an utility function) ? What do you think ?

Reply:
@francis0407 never mind .. i think it wouldn't be a InSubquery..

Reply:
@francis0407 sorry, the query is : 
select * from s where x in (select c1 from values (1), (2), (3) foo(c1));

Reply:
This should be a correlated predicate with a non-correlated subquery(or `ListQuery`). Note that `Exists` is a `SubqueryExpression` and `InSubquery` is a `Predicate`.

Reply:
@francis0407 Not sure what other reviewers think. I feel it would be better to move this method into the local scope (with probably a different name) of RewriteCorrelatedSubquery rather than keeping it as an utility function. However, i am fine if other reviewers are okay as personally i understand what you are trying to do here :-)

Reply:
I used to add this method in the Helper because this PR added an optimizer rule before (removed now), and it also used this method. Perhaps it is better to move it into the rule in this PR. Thanks @dilipbiswal.